keyboards/ergodox_ez/keymaps/dctucker/keymap.c

```diff
@@ -0,0 +1,344 @@
+	),
+	[4] = LAYOUT_ergodox_pretty(
+		_______, _______, _______, _______, _______, _______, _______,  _______, _______, _______, _______, _______, _______, _______,
+		_______, _______, _______, _______, _______, _______, _______,  _______, _______, _______, _______, _______, _______, _______,
+		KC_CAPS, _______, _______, _______, _______, _______,                    _______, _______, _______, _______, _______, _______,
+		_______, _______, _______, _______, _______, _______, _______,  _______, _______, _______, _______, _______, _______, _______,
+		_______, _______, _______, _______, _______,                                      _______, _______, _______, _______, _______,
+		                                             _______, _______,  _______, _______,
+		                                                      _______,  _______,
+		                                    _______, _______, _______,  _______, _______, _______
+	),
+};
+
+const uint16_t PROGMEM fn_actions[] = {
+  [1] = ACTION_LAYER_TAP_TOGGLE(1)
+};
+
+bool xtra_sent = true;
+bool ctl_xtra_sent = true;
+uint16_t lctl_timer, rctl_timer;
+bool caps = false;
+
+void keyboard_post_init_user(void) {
+	led_set_user(host_keyboard_leds());
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+	switch (keycode) {
+		// dynamically generate these.
+		case EPRM:
+			if (record->event.pressed) {
+				eeconfig_init();
+			}
+			return false;
+			break;
+		case RGB_SLD:
+			if (record->event.pressed) {
+				rgblight_mode(1);
+			}
+			return false;
+			break;
+		case CTLESC:
+			if( record->event.pressed ){
+				ctl_xtra_sent = false;
+				lctl_timer = timer_read();
+				register_code(KC_LCTL);
+			} else {
+				unregister_code(KC_LCTL);
+				if( ! ctl_xtra_sent ){
+					if( timer_elapsed(lctl_timer) < 500 ){
+						if( host_keyboard_leds() & (1<<USB_LED_CAPS_LOCK) ){
+							add_key(KC_RSFT);
+							add_key(KC_ESC);
+							send_keyboard_report();
+							del_key(KC_ESC);
+							del_key(KC_RSFT);
+							send_keyboard_report();
+						} else {
+							send_string("\033");
+						}
+					}
+					ctl_xtra_sent = true;
+				}
+			}
+			return false;
+			break;
+		case CTLENT:
+			if( record->event.pressed ){
+				ctl_xtra_sent = false;
+				rctl_timer = timer_read();
+				register_code(KC_RCTL);
+				//layer_on(4);
+			} else {
+				//layer_off(4);
+				unregister_code(KC_RCTL);
+				if( ! ctl_xtra_sent ){
+					if( timer_elapsed(rctl_timer) < 500 ){
+						send_string("\n");
+					}
+					ctl_xtra_sent = true;
+				}
+			}
+			return false;
+			break;
+		case RSHL4:
+			if( record->event.pressed ){
+				register_code(KC_RSFT);
+				layer_on(4);
+			} else {
+				layer_off(4);
+				unregister_code(KC_RSFT);
+			}
+			return false;
+			break;
+		case SPCAP:
+			if( record->event.pressed ){
+				register_code(KC_SPC);
+			} else {
+				unregister_code(KC_SPC);
+				if( caps ){ // if( host_keyboard_leds() & (1<<USB_LED_CAPS_LOCK) ){
```

Comment:
```suggestion
                if (host_keyboard_led_state().caps_lock) {
```

Reply:
I don't have receipts, but I'm remembering a situation where the indicators on the Ergodox didn't match the actual state of Caps lock that led to the decision to use a global state variable in this keymap.